gitpod-io/gitpod
# Add full clone setting for prebuilds

## Description

Improve large repository workspace startup performance by fully clone repository in prebuilds phase (download + unzip speed is faster than SCM processing)

## Related Issue(s)

Fixes ENT-730

## How to test

- Add `SUPERVISOR_DEBUG_ENABLE=true` to preview env's environments https://hw-large-repo.preview.gitpod-dev.com/user/variables so that we get debug logs from supervisor
- Create workspace with context https://github.com/JetBrains/intellij-community
   - 1-a) Check log of `ws-daemon` > `clone repository took`
   - 1-c) Check logs of workspace > `unshallow of local repository took`
- Create project for https://github.com/JetBrains/intellij-community, and enable prebuild (only enable), trigger prebuild and create a workspace after prebuild succeed
   - 2-a) Check logs of `ws-daemon`  > `clone repository took`
   - 2-b) Check logs of `ws-daemon` > `aria2c download duration`, `extract tarbal duration`
   - 2-c) Check logs of workspace > `unshallow of local repository took`
- Enable full clone of project in project settings page, trigger prebuild in UI and create a workspace after prebuild succeed
   - 3-a) Check logs of `ws-daemon`  > `clone repository took`
   - 3-b) Check logs of `ws-daemon` > `aria2c download duration`, `extract tarbal duration`
   - 3-c) Check logs of workspace > `unshallow of local repository took`
- Duration of `3-b + 3-c` should less than `1-a + 1-c` and `2-a + 2-b + 2-c`
- Workspace 3 git functions should works like workspace 1 /2 (i.e. `git lg`)

✅ Shared test result

| Duration | |Workspace 1|WS 2 | WS 3|
|:-:|:-:|:-:|:-:|:-:|
|||pure workspace|prebuild + workspace|prebuild(full clone) + workspace|
|**a)** |ws-daemon git clone|52.920714931s| `prebuild` 49.110767306s| `prebuild` 4m50.046374877s|
|**b)** | ws-daemon download and extract tar|NaN|8.946159792s+18.608708639s|26.825172768s+32.283299312s|
|||`accessable`|`accessable`|`accessable`|
|**c)** | supervisor fetch|7m35.843093833s|7m0.284969917s|2.52514ms|
|summary|| `access=53s`, `performance=7m35s`|`access=27.6s`, `performance=7m`|`access=59s`, `performance=0s`|

## Documentation


#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-large-repo</li>
	<li><b>🔗 URL</b> - <a href="https://hw-large-repo.preview.gitpod-dev.com/workspaces" target="_blank">hw-large-repo.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-large-repo-gha.28542</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-large-repo%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-dev-preview" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

      Run the build with werft instead of GHA
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

</details>

<details>
<summary>Installer</summary>

  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

      If enabled this will build `install/preview`
      If enabled this will create the environment on GCE infra
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
</details>

|Dashboard UI Changes|
|:-:|
|<img width="604" alt="image" src="https://github.com/user-attachments/assets/638fe67c-6d31-4d07-a156-5965a518a4a1">|

/hold
